### ISSUE TITLE
E-mail notifications stuck in a loop

### ISSUE THREAD

**Issue by guidovranken** [google/oss-fuzz#5253]
Every few hours a new e-mail is sent about each Cryptofuzz issue, see for example https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=30862

**Comment by davidben** [comment]
In addition to fixing this loop, can oss-fuzz switch to using bulk edits in Monorail so we only get one email when adding a CC, instead of one per issue? I haven't yet set an email filter for all oss-fuzz mail, but it's getting close to the point that I need one, which would kind of defeat the purpose. :-(

Better yet, does Monorail have a groups feature?

**Comment by guidovranken** [comment]
I support David's request.

Apart from that I am thinking of splitting the current project up and give each library its own OSS-Fuzz project.

- Much less "spam"/irrelevant bug reports for everyone (I know this is very annoying and I am very sorry that it happens)
- Probably much more effective. I initially set out from the idea that the more libraries I add to the same fuzzer, the more opportunities it has to detect discrepancies, but I think it is counter productive because fuzzer CPU cycles have to be spread over dozens of libraries.

I already do that for some libraries (nettle, relic, bearssl, wolfssl, libressl) and it works very well, with virtually no irrelevant reports. I compare these libraries only against Botan which supports virtually every operation that CF supports.

@davidben is that something you'd want with BoringSSL? We can create a new project like `boringssl-cryptofuzz` so I don't become privy to your other bug reports.

**Comment by guidovranken** [comment]
- A third advantage is that we can use i386 fuzzing again, which I disabled a long time ago because the current setup with dozens of libraries packed into a fuzzer consumes too much memory for i386 and makes it crash.

**Comment by davidben** [comment]
SGTM. Yeah, that sounds like a good improvement. Then I don't have to filter out bug reports for other libraries and the like.

One potential downside is we won't share test vectors across projects as easily. Like if OSS-Fuzz happened to find an interesting ECDSA input on one project, it'd be nice to run it on the others too. But hopefully that's a marginal benefit and something we can do at the OSS-Fuzz level rather than making jumbo fuzzers.

**Comment by guidovranken** [comment]
> SGTM. Yeah, that sounds like a good improvement. Then I don't have to filter out bug reports for other libraries and the like.
> 
> One potential downside is we won't share test vectors across projects as easily. Like if OSS-Fuzz happened to find an interesting ECDSA input on one project, it'd be nice to run it on the others too. But hopefully that's a marginal benefit and something we can do at the OSS-Fuzz level rather than making jumbo fuzzers.

Right, however I've invested a lot of energy in the CF mutator. It comes [preloaded with parameters for most curves](https://github.com/guidovranken/cryptofuzz/blob/61dd6983bcd12e7cb9a587aa93402a393a9823f1/gen_repository.py#L814-L1393) which it randomly inserts into ECDSA and other EC operations. It also reuses components, e.g. a ECDSA signature is reused as a public key etc. This frequently leads to interesting (though sometimes harmless) bugs like [this one](https://github.com/trezor/trezor-firmware/pull/1374). This is just to say that CF has become pretty good at finding ECDSA bugs and performs much better than just trying random mutations. There's also an option to import Wycheproof ECDSA test vectors into the CF corpus.

I will have a PR up in a minute.

**Comment by inferno-chromium** [comment]
I dont like the ideas of creating dozens of cryptofuzz-lib-name projects (unfair resource alloc). Can cryptofuzz be added to every existing crypto lib project in some way.

@jonathanmetzman @oliverchang @kcc  thoughts ?

**Comment by oliverchang** [comment]
+1 to bundling cryptofuzz with existing library projects, rather than cryptofuzz-lib-name projects. Is this possible?

**Comment by guidovranken** [comment]
> +1 to bundling cryptofuzz with existing library projects, rather than cryptofuzz-lib-name projects. Is this possible?

No problem, except that I will see BoringSSL's (and other project's) sensitive security bugs. Either that or I can exclude myself from the project, but then I won't be able to troubleshoot/debug issues found by my fuzzer.

**Comment by guidovranken** [comment]
To clarify, BoringSSL already runs a number of fuzzers that test their TLS and X509 implementations and some other things.

I wouldn't ever distribute bug reports but to the BoringSSL team I'm just an unvetted guy on the internet so making me privy to their vulnerabilities is a liability.

Additionally, if OSS-Fuzz finds a buffer overflow in their TLS stack today and tomorrow it's exploited in the wild, they will (understandably) think that I am distributing their bugs. Not having access to any information in the first place avoids this situation.

**Comment by guidovranken** [comment]
E-mail issue may be fixed with https://github.com/google/oss-fuzz/pull/5263
